GG/GG/DeferredLayout.h

```diff
@@ -0,0 +1,49 @@
+// -*- C++ -*-
+/** \file DeferredLayout.h \brief Contains the DefferedLayout class, which is used to size and
+    align GG windows in the PreRender() phase. */
+
+#ifndef _GG_DeferredLayout_h_
+#define _GG_DeferredLayout_h_
+
+#include <GG/Layout.h>
+
+namespace GG {
+
+/** \brief An invisible Wnd subclass that arranges its child Wnds during PreRender.
+
+    A DeferredLayout is a layout that does all of the expensive layout operations once per frame
+    during PreRender().
+*/
+class GG_API DeferredLayout : public Layout
+{
+public:
+    /** \name Structors */ ///@{
+    /** Ctor. */
+    DeferredLayout(X x, Y y, X w, Y h, std::size_t rows, std::size_t columns,
+                   unsigned int border_margin = 0, unsigned int cell_margin = INVALID_CELL_MARGIN);
+    //@}
+
+    /** \name Accessors */ ///@{
+    //@}
+
+    /** \name Mutators */ ///@{
+    virtual void SizeMove(const Pt& ul, const Pt& lr);
+    virtual void PreRender();
+    virtual void SetMinSize(const Pt& sz);
+    virtual void SetMaxSize(const Pt& sz);
+    //@}
+
+protected:
+    /** \name Mutators */ ///@{
+    virtual void RedoLayout();
```

Comment:
I may not be understanding your question, so if the answer misses your point I apologize.

Three facts are significant:

1. `Layout::RedoLayout()` and `DeferredLayout::RedoLayout()` do different things.

 `Layout::RedoLayout()` immediately causes a re-layout to happen if the size has changed.

 `DeferredLayout::RedoLayout()` calls `RequirePreRender()` which will cause a `DoLayout()` within `PreRender()`.


2. `Layout::RedoLayout()` is a virtual function.

 Since `RedoLayout()` is a virtual function every place in `Layout::SomeMutatorFunction()` that calls `RedoLayout()` will call `DeferredLayout::RedoLayout()` if the `Layout` is really a `DeferredLayout`.  This makes the changes in all of the mutator function in `Layout` cause deferred changes if the `Layout` is really a `DeferredLayout`.

3. `Layout::RedoLayout()` is protected because only `Layout` or its derived classes will know when a change should trigger a `RedoLayout()`.  Of course, anyone can call the public `RequirePreRender()`, but that doesn't do anything for `Layout`.